pkg/cmd/provisioning/gcp/create_workload_identity_provider.go

```diff
@@ -0,0 +1,252 @@
+		}
+	}
+
+	return nil
+}
+
+func createOIDCConfiguration(ctx context.Context, client gcp.Client, bucketName, issuerURL, targetDir string, generateOnly bool) error {
+	discoveryDocumentJSON := provisioning.GenerateDiscoveryDocument(issuerURL)
+	if generateOnly {
+		err := provisioning.SaveToFile("discovery document", filepath.Join(targetDir, provisioning.GCPOidcConfigurationFilename), []byte(discoveryDocumentJSON))
+		if err != nil {
+			return errors.Wrap(err, "failed to save oidc configuration file")
+		}
+	} else {
+		err := client.PutObject(ctx, bucketName, provisioning.DiscoveryDocumentURI, []byte(discoveryDocumentJSON))
+		if err != nil {
+			return errors.Wrapf(err, "failed to upload discovery document in the bucket %s", bucketName)
+		}
+		log.Printf("OpenID Connect discovery document in the S3 bucket %s at %s updated", bucketName, provisioning.DiscoveryDocumentURI)
+	}
+	return nil
+}
+
+func createJSONWebKeySet(ctx context.Context, client gcp.Client, publicKeyFilepath, bucketName, targetDir string, generateOnly bool) error {
+	jwks, err := provisioning.BuildJsonWebKeySet(publicKeyFilepath)
+	if err != nil {
+		return errors.Wrap(err, "failed to build JSON web key set from the public key")
+	}
+
+	if generateOnly {
+		err = provisioning.SaveToFile("JSON web key set (JWKS)", filepath.Join(targetDir, provisioning.GCPOidcKeysFilename), jwks)
+		if err != nil {
+			return errors.Wrap(err, "failed to save keys.json file")
+		}
+	} else {
+		err = client.PutObject(ctx, bucketName, provisioning.DiscoveryDocumentURI, jwks)
+		if err != nil {
+			return errors.Wrapf(err, "failed to upload JSON web key set (JWKS) in the S3 bucket %s", bucketName)
+		}
+		log.Printf("JSON web key set (JWKS) in the S3 bucket %s at %s updated", bucketName, provisioning.KeysURI)
+	}
+	return nil
+}
+
+func createIdentityProvider(ctx context.Context, client gcp.Client, name, displayName, description, project, issuerURL, workloadIdentityPool, targetDir string, generateOnly bool) error {
+	if generateOnly {
+		createIdentityProviderScriptFilepath := filepath.Join(targetDir, provisioning.CreateIdentityProviderScriptName)
+		script := fmt.Sprintf(createIdentityProviderScript, name, workloadIdentityPool, displayName, description, issuerURL, provisioning.OpenShiftAudience)
+		err := provisioning.SaveToFile("shell script to create workload identity provider", createIdentityProviderScriptFilepath, []byte(script))
+		if err != nil {
+			return errors.Wrap(err, "failed to save shell script to create workload identity provider")
+		}
+	} else {
+		provider := &iam.WorkloadIdentityPoolProvider{
+			Name:        name,
+			DisplayName: displayName,
+			Description: description,
+			State:       "ACTIVE",
+			Disabled:    false,
+			Oidc: &iam.Oidc{
+				AllowedAudiences: []string{provisioning.OpenShiftAudience},
+				IssuerUri:        issuerURL,
+			},
+			AttributeMapping: map[string]string{
+				"google.subject": "assertion.sub",
+			},
```

Comment:
"google.subject": "assertion.sub" only maps whatever is there in oidc token sub into google auth token sub when token exchange happens. Currently oidc token sub is something like `system:serviceaccount:<sa_namespace>:sa_name`. Do you think provider information in sub is required? if yes, what purpose does it serve?

Reply:
Since there will be only one provider in the pool, going ahead with the google [recommendation](https://cloud.google.com/iam/docs/manage-workload-identity-pools-providers) `"google.subject": "assertion.sub"`

For multiple providers google suggests following, but there is hard limit of 127 charactcers
`google.subject="provider-a::" + assertion.sub
`

Reply:
:+1: The assertion.sub seems to map with the AWS pattern of associating an IAM Role via `[OIDC-name]:sub` with the ServiceAccount(s).